lightningd/bitcoind.c

```diff
@@ -783,6 +783,85 @@ void bitcoind_gettxout(struct bitcoind *bitcoind,
 			  NULL);
 }
 
+static bool extract_numeric_version(struct bitcoin_cli *bcli,
+			    const char *output, size_t output_bytes,
+			    u64 *version)
+{
+	const jsmntok_t *tokens, *versiontok;
+	bool valid;
+
+	tokens = json_parse_input(output, output, output_bytes, &valid);
+	if (!tokens)
+		fatal("%s: %s response",
+		      bcli_args(tmpctx, bcli),
+		      valid ? "partial" : "invalid");
+
+	if (tokens[0].type != JSMN_OBJECT) {
+		log_unusual(bcli->bitcoind->log,
+			    "%s: gave non-object (%.*s)?",
+			    bcli_args(tmpctx, bcli),
+			    (int)output_bytes, output);
+		return false;
+	}
+
+	versiontok = json_get_member(output, tokens, "version");
+	if (!versiontok)
+		return false;
+
+	return json_to_u64(output, versiontok, version);
+}
+
+/* Supported version is the prerequisite for normal runnning.
+ * So any error will call `fatal` and fail the `lightningd`.
+ */
+static bool process_getclientversion(struct bitcoin_cli *bcli)
+{
+	u64 version;
+	u64 min_version = bcli->bitcoind->chainparams->cli_min_supported_version;
+
+	if (*bcli->exitstatus != 0) {
+		/* bitcoin/src/rpc/protocol.h:
+		 * 	RPC_METHOD_NOT_FOUND = -32601
+		 */
+		if (*bcli->exitstatus != 32601)
```

Comment:
emm, it should be `89`. We use `WEXITSTATUS` to return low-order 8 bits of the exit status value. For `32601`, the corresponding exitstatus is `89`.(I not sure why it is a positive number).
And I confirmed again, from v0.14.0, `getnetworkinfo` was available, and `getinfo` was deprecated and was removed in v0.16.0. So I should use `getnetworkinfo` here.

Reply:
What about `fatal` for the case if `exitstatus==89`(this means the version older than v0.14), and call `bcli_failure` for other non-zero exitstatus?

Reply:
Would it work to just try `getinfo` and if that fails (regardless of failure reason) then `getnetworkinfo`?

Reply:
Yes, it will work. But there are much more people runs the newer(than v0.14.0) version, and the `getinfo` has the very high probability to failure. Maybe we call `getnetworkinfo` first is better?
If `89` is ugly, we can remove  this `exitstatus==89` code block, and just (use `getnetworkinfo`) call `bcli_failure` for all non-zero exitstatus.
Thus, we just set `nonzero_exit_ok` parameter as `false` when we call `start_bitcoin_cli`, and any non-zero exitstatus will trigger `bcli_failure`.
What do you think about the code now?

Reply:
Looks ok.